File: 코드리뷰/장문주(jmj2359)/10일차/코드리뷰_치즈.py
import sys
sys.setrecursionlimit(100000)
n, m = map(int, sys.stdin.readline().rstrip().split())

dx = [1,0,-1,0]
dy = [0,-1,0,1]

arr = []
for i in range(n):
    arr+=[list(map(int,sys.stdin.readline().rstrip().split()))]

def dfs(x,y):
    visited[x][y] = -1
    for i in range(4):
        newx = x+dx[i]
        newy = y+dy[i]
        if (newx < 0) or (newy < 0) or (newx > n-1) or (newy > m-1):
            continue
        if (arr[newx][newy] == 1):
            visited[newx][newy]+=1
        if visited[newx][newy] == False and arr[newx][newy] == 0:
            dfs(newx,newy)

cheese = 0
for i in range(n):
    for j in range(m):
        if arr[i][j] >= 1:
            cheese += 1
            
            
ans = 0
while (cheese > 0):
    ans+=1
    visited = [[0 for i in range(m)] for j in range(n)]
    dfs(0,0)
    for i in range(n):
        for j in range(m):
            if (arr[i][j] == 1 and visited[i][j] >= 2):
                arr[i][j] = 0
                cheese-=1

print(ans)
    

File: 코드리뷰/이세영(0112leesy)/10일차/코드리뷰_치즈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class main {
    static int[][] arr;
    static boolean[][] visited;
    static int[] dx = {-1,0,1,0};
    static int[] dy = {0,-1,0,1};
    static int n,m;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        visited = new boolean[n][m];
        int cheese = 0;
        for(int i=0; i<n; i++) {
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<m; j++) {
                int num = Integer.parseInt(st.nextToken());
                if(num==1) cheese++;
                arr[i][j] = num;
            }
        }

        int ans = 0;
        while(cheese > 0) {
            ans++;
            visited = new boolean[n][m];
            dfs(0,0);
            for(int i=0; i<n; i++) {
                for(int j=0; j<m; j++) {
                    if(arr[i][j] == 1) {
                        int cnt = 0;
                        for(int k=0; k<4; k++) {
                            int new_i = i+dx[k];
                            int new_j = j+dy[k];
                            if(arr[new_i][new_j] == 0 && visited[new_i][new_j]) cnt++;
                        }
                        if(cnt >= 2) {
                            arr[i][j] = -1;
                            cheese--;
                        }
                    }
                }
            }
            
            for(int i=0; i<n;i++) {
                for(int j=0; j<m; j++) {
                    if(arr[i][j] == -1) {
                        arr[i][j] = 0;
                    }
                }
            }            
        }
        System.out.println(ans);
    }
    
    static void dfs(int x, int y) {
        visited[x][y] = true;
        for(int i=0; i<4; i++) {
            int new_x = x+dx[i];
            int new_y = y+dy[i];
            if(new_x<0 || new_y<0 || new_x>=n || new_y>=m) continue;
            if(!visited[new_x][new_y] && arr[new_x][new_y]==0) {
                dfs(new_x, new_y);
            }
        }
    }

}
